src/test/txpackage_tests.cpp

```diff
@@ -73,15 +122,243 @@ BOOST_FIXTURE_TEST_CASE(package_sanitization_tests, TestChain100Setup)
+        // in package.
+        for (auto i{0}; i < 99; ++i) {
+            BOOST_CHECK_EQUAL(packageified.FilteredAncestorSet(package[i])->size(), 1);
+            BOOST_CHECK(IsAncestorPackage(*packageified.FilteredAncestorSet(package[i])));
+            if (i < 50) packageified.Skip(package[i]);
+        }
+        // After excluding 50 of the parents, the child's ancestor set has size 50.
+        BOOST_CHECK_EQUAL(packageified.FilteredAncestorSet(package.back())->size(), 50);
+        BOOST_CHECK(IsAncestorPackage(*packageified.FilteredAncestorSet(package.back())));
+        packageified.SkipWithDescendants(package[75]);
+        for (auto i{50}; i < 99; ++i) {
+            if (i == 75) {
+                BOOST_CHECK(!packageified.FilteredAncestorSet(package[i]));
+            } else {
+                BOOST_CHECK_EQUAL(packageified.FilteredAncestorSet(package[i])->size(), 1);
+            }
+        }
+        BOOST_CHECK(!packageified.FilteredAncestorSet(package.back()));
+    }
+
+    // Heavily inter-connected set of 50 transactions
+    LOCK2(cs_main, m_node.mempool->cs);
+    auto transactions{PopulateMempool(det_rand, /*num_transactions=*/50, /*submit=*/true)};
+    Shuffle(transactions.begin(), transactions.end(), det_rand);
+    AncestorPackage packageified{transactions};
+    const auto sorted_transactions{packageified.Txns()};
+    BOOST_CHECK(IsSorted(sorted_transactions));
+    for (const auto& tx : sorted_transactions) {
+        const auto packageified_ancestors{packageified.FilteredAncestorSet(tx)};
+        BOOST_CHECK(IsAncestorPackage(*packageified_ancestors));
+        auto mempool_ancestors{m_node.mempool->CalculateMemPoolAncestors(*m_node.mempool->GetIter(tx->GetHash()).value(),
+                               CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false)};
+        // Add 1 because CMPA doesn't include the tx itself in its ancestor set.
+        BOOST_CHECK_EQUAL(mempool_ancestors->size() + 1, packageified_ancestors->size());
+        std::set<uint256> packageified_ancestors_wtxids;
+        for (const auto& tx : packageified_ancestors.value()) packageified_ancestors_wtxids.insert(tx->GetWitnessHash());
+        for (const auto& mempool_iter : *mempool_ancestors) {
+            BOOST_CHECK(packageified_ancestors_wtxids.count(mempool_iter->GetTx().GetWitnessHash()) > 0);
+        }
+    }
+    // Skip the 20th transaction. All of its descendants should have 1 fewer tx in their ancestor sets.
+    const auto& tx_20{sorted_transactions[20]};
+    CTxMemPool::setEntries descendants_20;
+    m_node.mempool->CalculateDescendants(m_node.mempool->GetIter(tx_20->GetHash()).value(), descendants_20);
+    packageified.Skip(tx_20);
+    for (const auto& desc_iter : descendants_20) {
+        BOOST_CHECK_EQUAL(packageified.FilteredAncestorSet(m_node.mempool->info(GenTxid::Txid(desc_iter->GetTx().GetHash())).tx)->size(),
```

Comment:
You should check if `FilteredAncestorSet` is giving back a nullopt or not, otherwise it's UB to call `->Size()` on it when it's nullopt